subsys/sip_svc/sip_svc_shell.c

```diff
@@ -0,0 +1,309 @@
+{
+	struct sip_svc_controller *ctrl;
+	uint64_t c_token;
+	int err;
+	char *ptr;
+
+	err = parse_common_args(sh, argv, &ctrl);
+	if (err < 0) {
+		return err;
+	}
+
+	c_token = strtol(argv[2], &ptr, 16);
+
+	err = sip_svc_unregister(ctrl, (uint32_t)c_token);
+	if (err) {
+		shell_error(sh, "%s: unregister fail (%d): client token %08x",
+			    ctrl->method, err, (uint32_t)c_token);
+	} else {
+		shell_print(sh, "%s: unregister success: client token %08x",
+			    ctrl->method, (uint32_t)c_token);
+	}
+
+	return err;
+}
+
+static int cmd_open(const struct shell *sh, size_t argc, char **argv)
+{
+	struct sip_svc_controller *ctrl;
+	uint64_t c_token;
+	uint32_t usecond = 0;
+	int err;
+	char *ptr;
+
+	err = parse_common_args(sh, argv, &ctrl);
+	if (err < 0) {
+		return err;
+	}
+
+	c_token = strtol(argv[2], &ptr, 16);
+
+	if (argc > 3) {
+		usecond = (uint32_t)atoi(argv[3]) * 1000000;
+	}
+
+	err = sip_svc_open(ctrl, (uint32_t)c_token, usecond);
+	if (err) {
+		shell_error(sh, "%s: open fail (%d): client token %08x",
+			    ctrl->method, err, (uint32_t)c_token);
+	} else {
+		shell_print(sh, "%s: open success: client token %08x",
+			    ctrl->method, (uint32_t)c_token);
+	}
+
+	return err;
+}
+
+static int cmd_close(const struct shell *sh, size_t argc, char **argv)
+{
+	struct sip_svc_controller *ctrl;
+	uint64_t c_token;
+	int err;
+	char *ptr;
+
+	err = parse_common_args(sh, argv, &ctrl);
+	if (err < 0) {
+		return err;
+	}
+
+	c_token = strtol(argv[2], &ptr, 16);
+
+	err = sip_svc_close(ctrl, (uint32_t)c_token);
+	if (err) {
+		shell_error(sh, "%s: close fail (%d): client token %08x",
+			    ctrl->method, err, (uint32_t)c_token);
+	} else {
+		shell_print(sh, "%s: close success: client token %08x",
+			    ctrl->method, (uint32_t)c_token);
+	}
+
+	return err;
+}
+
+static void cmd_send_callback(uint32_t c_token, char *res, int size)
+{
+	struct sip_svc_response *response =
+		(struct sip_svc_response *) res;
+
+	struct private_data *priv = (struct private_data *)response->priv_data;
+	const struct shell *sh = priv->sh;
+
+	shell_print(sh, "\n\rsip_svc send callback response\n");
+
+	if ((size_t)size != sizeof(struct sip_svc_response)) {
+		shell_error(sh, "Invalid response size (%d), expects (%ld)",
+			size, sizeof(struct sip_svc_response));
+		return;
+	}
+
+	shell_print(sh, "\theader=%08x\n", response->header);
+	shell_print(sh, "\ta0=%016lx\n", response->a0);
+	shell_print(sh, "\ta1=%016lx\n", response->a1);
+	shell_print(sh, "\ta2=%016lx\n", response->a2);
+	shell_print(sh, "\ta3=%016lx\n", response->a3);
+
+	k_sem_give(&(priv->semaphore));
+}
+
+static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+{
+	struct sip_svc_controller *ctrl;
+	uint64_t c_token;
+	int trans_id;
+	struct sip_svc_request request;
+	struct private_data *priv = NULL;
+	int err;
+	char *ptr;
+
+	err = parse_common_args(sh, argv, &ctrl);
+	if (err < 0)
+		return err;
+
+	priv = (struct private_data *)k_malloc(sizeof(struct private_data));
```

Comment:
Why use the heap? Put it on the stack and pass a pointer to the callback function?